docs/dev/data_validation.rst

```diff
@@ -0,0 +1,688 @@
+
+.. warning::
+
+  You should rarely if ever need to edit the row-counts file directly. It needs to be
+  kept sorted to minimize diffs in git, and manually calculating and editing row counts
+  is both tedious and error prone.
+
+Checking for entirely null columns
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The other test we apply to basically all tables is ``expect_columns_not_all_null``. In
+its most basic form it verifies that there are no columns in the table which are
+completely null, since that is typically indicative of a bad ``ENUM`` constraint, a
+column naming error, or a bad merge, and should be investigated. To add this basic
+default, you add the test to the table level ``data_tests`` with no parameters, which
+building on the above example would look like:
+
+.. code-block:: yaml
+
+    version: 2
+    sources:
+      - name: pudl
+        tables:
+          - name: new_table_name
+            data_tests:
+              - expect_columns_not_all_null
+              - check_row_counts_per_partition:
+                  table_name: new_table_name
+                  partition_column: report_date
+
+--------------------------------------------------------------------------------
+Defining new data validation tests
+--------------------------------------------------------------------------------
+
+* How to define a new generic test (lean on references to dbt docs when possible)
+* Focus on the things that make the PUDL use case unusual.
+* DuckDB + Parquet means we can't rely on ``adapter`` object methods (no real DB)
+* Almost all our tables are "sources" not "models"
+
+Defining Macros
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* In dbt, macros are reusable SQL snippets that can be used to simplify your tests. You
+  can define a macro once and then use it in multiple tests. This is particularly useful
+  for complex tests that require a lot of boilerplate code.
+
+Testing the Tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* One reason to create macros for more complex functions is that they can be
+  independently unit-tested.
+
+Creating intermediate tables for a test
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. todo::
+
+   Do we actually have any tests that do the first option below? I don't really
+   understand what it is suggesting :user:`zaneselvans`
+
+In some cases you may need to modify a table or calculate some derived values before
+you can apply a test. There are two ways to accomplish this. First, you can add the
+table as a ``source`` as described above, then create a SQL file in the ``tests/``
+directory like ``tests/{data_source}/{table_name}.yml``.  From here you can construct a
+SQL query to modify the table and execute a test on the intermediate table you've
+created. ``dbt`` expects a SQL test to be a query that returns 0 rows for a successful
+test. See the ``dbt`` `source function
+<https://docs.getdbt.com/reference/dbt-jinja-functions/source>`_ for guidance on how to
+reference a ``source`` from a SQL file.
```

Comment:
This paragraph didn't really make sense to me, or seem like something we were actually doing, but maybe I'm missing something?